### DISCUSSION TITLE
Implementing CORS to browser when authentication is required

### DISCUSSION THREAD

**Discussion by SampsaKaskela** [AxisCommunications/discussions#223] [VAPIX]
We have web application that displays camera. We are using `media-stream-library` to do this. But before using that we are making request to `usergroup.cgi` endpoint to authenticate the user. This works fine if in browser Fetch API we are putting in `no-cors` option and credentials `include` option. But we also want user to be able to view camera recordings so we want to make call to `record/list.cgi`. But we can't use no-cors option in this case because we want to receive response body. We then used `customhttpheader.cgi` to configure CORS headers. We set following headers:

```
Access-Control-Allow-Origin: http://localhost:3003 (this is for dev purposes only)
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: authorization
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
```

With no authentication (anonymous viewing enabled) CORS now works and we can connect to API and to camera. But we don't want camera to be accessible by everyone so we want to require authentication. But when authentication is enabled and browser tries to make request we get CORS error due to preflight request (method OPTIONS) getting 401 Unauthorized response. To our knowledge OPTIONS request should always return 200 OK and then CORS will work correctly. Currently we can't get this to work due to this issue and we don't know what is the problem. We could proxy the request via our own server to avoid CORS but this would increase our load to server so we would prefer to be able to connect to camera directly from browser.

is this a bug or are we missing some setting in our camera that would make this work?

Camera: AXIS P3245-VE
Firmware: 11.2.68

**Comment by DavGarcia** [comment]
Did you find a solution for this issue? We are facing the same.

**Reply by SampsaKaskela** [reply]
No solution so far. I have tried to contact Axis directly for solution since our company is partner but so far no response. I'm sure to update here when I receive response.

**Comment by slgtech** [comment]
How did you set the CORS option on the camera? I am trying to set the Access-Control-Allow-Origin using the VAPIX command with no success

**Reply by SampsaKaskela** [reply]
> How did you set the CORS option on the camera? I am trying to set the Access-Control-Allow-Origin using the VAPIX command with no success

The camera uses websocket so no need to setup CORS there. media-stream-library has examples how to authenticate using "no-cors" option. Basically just make request to any other endpoint before connecting with websocket so you get credentials from server. If you just want the camera then you don't need CORS.

Our issue here is to access VAPIX endpoint with CORS when authentication is required. You can see my configuration above which works without authentication but fails with authentication.

**Comment by GuYounes** [comment]
Why is this discussion marked as answered ?

**Reply by vivekatoffice** [reply]
Hi @GuYounes ,
I've reopened the discussion for you. Feel free to share more details about the CORS issue you're encountering.
Alternatively, you're welcome to start a new discussion if that's more convenient.

**Comment by SampsaKaskela** [comment]
This turned out to be bug in Axis software and this should be fixed in latest firmware. Requests should now work with at least at endpoints that I mentioned although haven't tested it for all the endpoints. So this should technically be solved and if you have older firmware then you need to update or if you can't implement proxy server that handles things correctly.

**Comment by vivekatoffice** [comment]
Feel free to re-open it OR share the solution/alternative/comment if already resolved👁️‍🗨️
